### PR TITLE
add CompositeDictionary.

### PR DESCRIPTION
**Pull request by vwxyzh**
@chenkennt @vicancy @ansyral @hellosnow @qinezh @superyyrrzz

### CODE REVIEW

File: src/Microsoft.DocAsCode.Common/CompositeDictionary.cs [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+                }
+                foreach (var entry in _entries)
+                {
+                    if (key.StartsWith(entry.Prefix))
+                    {
+                        entry.SetValue(key.Substring(entry.Prefix.Length), value);
+                        return;
+                    }
+                }
+            }
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Common/CompositeDictionary.cs]: also throw new KeyNotFoundException() if not found?

Comment (hellosnow) [inline src/Microsoft.DocAsCode.Common/CompositeDictionary.cs]: I mean whether should set the value when the key not starts with prefix?

---
In reply to: [97214579](https://github.com/dotnet/docfx/pull/1216#discussion_r97214579) [](ancestors = 97214579)

File: src/Microsoft.DocAsCode.Common/CompositeDictionary.cs [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+                foreach (var entry in _entries)
+                {
+                    if (key.StartsWith(entry.Prefix))
+                    {
+                        var pair = entry.TryGetValue(key.Substring(entry.Prefix.Length));
+                        if (pair.Key)
+                        {
+                            return pair.Value;
+                        }
+                        break;
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Common/CompositeDictionary.cs]: >break [](start = 24, length = 5)

remove?

**Comment by hellosnow** [comment]
:shipit:

File: src/Microsoft.DocAsCode.Common/CompositeDictionary.cs [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+                }
+                if (dict == null)
+                {
+                    throw new ArgumentNullException(nameof(dict));
+                }
+                if (valueConverter == null)
+                {
+                    valueConverter = o => (TValue)o;
+                }
+                _entries.Add(new Entry
```
Comment (ansyral) [inline src/Microsoft.DocAsCode.Common/CompositeDictionary.cs]: >_entries.Add [](start = 16, length = 12)

check whether the prefix already exists?

**Comment by ansyral** [comment]
:shipit:

**Comment by vicancy** [comment]
:shipit:

**Pull request merged**